pkgs/build-support/docker/stream_layered_image.py

```diff
@@ -0,0 +1,250 @@
+"""
+This script generates a Docker image from a set of store paths. Uses
+Docker Image Specification v1.2 as reference [1].
+
+It expects a JSON file with the following properties and writes the
+image as an uncompressed tarball to stdout:
+
+* "architecture", "config", "os", "created", "repo_tag" correspond to
+  the fields with the same name on the image spec [2].
+* "created" can be "now".
+* "created" is also used as mtime for files added to the image.
+* "store_layers" is a list of layers in ascending order, where each
+  layer is the list of store paths to include in that layer.
+
+The main challenge for this script to create the final image in a
+streaming fashion, without dumping any intermediate data to disk
+for performance.
+
+A docker image has each layer contents archived as separate tarballs,
+and they later all get enveloped into a single big tarball in a
+content addressed fashion. However, because how "tar" format works,
+we have to know about the name (which includes the checksum in our
+case) and the size of the tarball before we can start adding it to the
+outer tarball.  We achieve that by creating the layer tarballs twice;
+on the first iteration we calculate the file size and the checksum,
+and on the second one we actually stream the contents. 'add_layer_dir'
+function does all this.
+
+[1]: https://github.com/moby/moby/blob/master/image/spec/v1.2.md
+[2]: https://github.com/moby/moby/blob/4fb59c20a4fb54f944fe170d0ff1d00eb4a24d6f/image/spec/v1.2.md#image-json-field-descriptions
+"""  # noqa: E501
+
+
+import io
+import os
+import re
+import sys
+import json
+import hashlib
+import tarfile
+import itertools
+import threading
+from datetime import datetime
+from collections import namedtuple
+
+
+# Adds the given store paths to as a tar to the given writable stream.
+def archive_paths_to(obj, paths, mtime, add_nix, filter=None):
```

Comment:
WDYT about adding docstrings to all Python methods?

```python
def archive_paths_to(obj, paths, mtime, add_nix, filter=None):
    """
    Adds the given store paths something something tar file something stream

    obj: something
    paths: something
    add_nix: something
    filter: something
    """
```
"""

the whole thing isn't super obvious to grasp (for me anyway) so i'd appreciate that ;)

```